gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+)
+
+// Transforms is a list of Transformer objects.
+// Transformer is the interface that wraps the basic Transform method.
+type TransformChain struct {
+	transforms []Transform
+	inputState *api.GlobalState
+}
+
+func CreateTransformChain(inputState *api.GlobalState, transforms []Transform) TransformChain {
+	return TransformChain{
+		transforms: transforms,
+		inputState: inputState,
+	}
+}
+
+func (chain *TransformChain) handleInitialState(state *api.GlobalState) (*api.GlobalState, error) {
+	// Melih TODO: implement
+	return state, nil
+}
+
+// Melih TODO: Check with Alan
+// command ID? Seems to be only used in logging and error
+// Should we log both maybe? i.e 5:4 [5th command generated 4th command]
+// Is it the state mutation to do?
+func (chain *TransformChain) mutateState(ctx context.Context, id api.CmdID, cmds []api.Cmd, state *api.GlobalState) error {
+	for i, cmd := range cmds {
+		if err := cmd.Mutate(ctx, id, state, nil, nil); err != nil {
+			log.W(ctx, "State mutation error in command [%v:%v]:%v", id, i, cmd)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (chain *TransformChain) BeginChain(ctx context.Context) ([]api.Cmd, error) {
+	chain.handleInitialState(chain.inputState)
+
+	cmds := make([]api.Cmd, 0)
+	var err error
+	for _, transform := range chain.transforms {
+		cmds, err = transform.BeginTransform(ctx, cmds, chain.inputState)
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+			// return nil, err
+		}
+
+		err = chain.mutateState(ctx, 0, cmds, chain.inputState)
```

Comment:
I know you haven't done the handleInitialState() / needsAccurateState() work yet, but this should probably be outside the loop until you handle that stuff. When that stuff is implemented, this will be inside the loop, but there will be something like "if transform.needsAccurateState() or isLastTransform() {" guarding it, and the state that is mutated will be selected based on the transform for this loop iteration.